### PR TITLE
feat: handle creating post moderation when required

### PR DESCRIPTION
**Pull request by AmarTrebinjac**
## Changes

Created logic for adding a post to moderation instead of creating it when moderation is required

## Events

Did you introduce any new tracking events?

## Experiment

Did you introduce any new experiments?

## Manual Testing

> [!CAUTION]
> Please make sure existing components are not breaking/affected by this PR

MI-608 #done
MI-583 #done

### Preview domain
https://mi-608.preview.app.daily.dev

### CODE REVIEW

File: packages/shared/src/components/modals/post/CreateSharedPostModal.tsx
Comment (rebelchris) [inline packages/shared/src/components/modals/post/CreateSharedPostModal.tsx:113]: Why are these 2 separate states.
From user perspective they just adding a new post right, don't think you need 2 endpoints for it.

**Review by rebelchris — COMMENTED** [review]
Should the frontend really handle the case of creation though?
For draft etc it's fine, but the initial one i'm not sure.

File: packages/shared/src/components/modals/post/CreateSharedPostModal.tsx
Comment (AmarTrebinjac) [reply packages/shared/src/components/modals/post/CreateSharedPostModal.tsx:113]: @rebelchris For the user the only difference will be the seeing the post vs seeing the pending approval modal.

As for to why there's two endpoints, we had a discussion on it here:
https://github.com/dailydotdev/daily-api/pull/2373#discussion_r1824057034

Comment (AmarTrebinjac) [reply packages/shared/src/components/modals/post/CreateSharedPostModal.tsx]: @rebelchris I think you were right and did a rework to more match your suggestion. 

However, I still left the creation of freeform post more or less how it is, since it didn't have its own hook and is handled in the page. I can create one if that is desired, though.

**Comment by AmarTrebinjac** [comment]
Merged [MI-583](https://dailydotdev.atlassian.net/jira/software/projects/MI/boards/31?selectedIssue=MI-583) into this because it's such a small task, and made it easier to test flow.

File: packages/shared/src/components/modals/post/CreateSharedPostModal.tsx
Comment (rebelchris) [inline packages/shared/src/components/modals/post/CreateSharedPostModal.tsx:68]: Maybe it's better if the `onSubmitPost` has a callback signature?
That way we can avoid the useEffect here.

File: packages/shared/src/components/squads/utils.tsx
Comment (rebelchris) [inline packages/shared/src/components/squads/utils.tsx:52]: This is not very well named I guess.
It's privileged member but specifically for moderatePost?

Either we should change the function to include `SourcePermissions` as secondary prop (preferred I think)
Or we change the name to `hasModeratePostPermissions`

File: packages/shared/src/graphql/posts.ts
Comment (rebelchris) [inline packages/shared/src/graphql/posts.ts:562]: Why are there 2 new types?

File: packages/shared/src/hooks/source/useSourcePostModeration.ts
Comment (rebelchris) [inline packages/shared/src/hooks/source/useSourcePostModeration.ts:46]: Rather use onSettled or something like that maybe?

File: packages/webapp/pages/squads/create.tsx
Comment (rebelchris) [inline packages/webapp/pages/squads/create.tsx:101]: Is this the only place we call it?
Wasn't sure if we had 2 ways to create freeform still (like a modal or something)

I do think ideally I'd prefer a hook (maybe even the same one) for all cases.
but don't want to block too much.

**Review by rebelchris — CHANGES_REQUESTED** [review]
Some minor comments.

File: packages/shared/src/hooks/source/useSourcePostModeration.ts
Comment (AmarTrebinjac) [reply packages/shared/src/hooks/source/useSourcePostModeration.ts:46]: Great suggestion!

File: packages/shared/src/components/squads/utils.tsx
Comment (AmarTrebinjac) [reply packages/shared/src/components/squads/utils.tsx]: Found we had the `verifyPermission` function which does what you describe, so I removed `isPrivilegedMember` and use that instead

File: packages/shared/src/components/modals/post/CreateSharedPostModal.tsx
Comment (AmarTrebinjac) [reply packages/shared/src/components/modals/post/CreateSharedPostModal.tsx]: I added an onSettled here as well

File: packages/webapp/pages/squads/create.tsx
Comment (AmarTrebinjac) [reply packages/webapp/pages/squads/create.tsx:101]: It is indeed the only place where createPost, which is the only place that CREATE_POST_MUTATION is referenced.

I agree that we could move it into the usePostToSquad hook, and we should, but like you said, would like to close off the remaining tasks for this RFC, so perhaps a task for another time ? :)

File: packages/shared/src/hooks/source/useSourcePostModeration.ts
Comment (sshanzel) [inline packages/shared/src/hooks/source/useSourcePostModeration.ts:41]:
```suggestion
    onSuccess,
    onError,
```

File: packages/shared/src/hooks/usePrompt.ts
Comment (sshanzel) [inline packages/shared/src/hooks/usePrompt.ts:15]:
```suggestion
  icon?: ReactNode;
```

File: packages/webapp/pages/squads/create.tsx
Comment (sshanzel) [reply packages/webapp/pages/squads/create.tsx:101]: If so, let's maybe write a ticket for this so we can come back when we get more time.

File: packages/shared/src/hooks/squads/usePostToSquad.ts
Comment (sshanzel) [inline packages/shared/src/hooks/squads/usePostToSquad.ts:184]: This format is for submitting external link, the parameters should not need the `externalLink`.

**Review by sshanzel — CHANGES_REQUESTED** [review]
We are almost ready! Just noticed one thing about sharing an existing post.

**Review by sshanzel — APPROVED** [review]
Don't have anything blocking! Looks amazing to me! Let's just wait for Luca to have a final look, then I think we can merge 🚀

**Review by ilasw — CHANGES_REQUESTED** [review]
Tested e2e and looks good, just 2 points.

File: packages/shared/src/hooks/source/useSourcePostModeration.ts
Comment (AmarTrebinjac) [reply packages/shared/src/hooks/source/useSourcePostModeration.ts:40]: @ilasw Initially I had an useEffect that depended on the "isSuccess" state. Do you think that would work better in this case?

**Review by ilasw — APPROVED** [review]
Great work 🚀

**Pull request merged**
